third-party/realdds/src/dds-metadata-syncer.cpp

```diff
@@ -8,73 +8,104 @@
 namespace realdds {
 
 
+dds_metadata_syncer::dds_metadata_syncer() :
+    _is_alive( std::make_shared< bool >( true ) )
+{
+}
+
+
+dds_metadata_syncer::~dds_metadata_syncer()
+{
+    *_is_alive = false;
+
+    std::lock_guard< std::mutex > lock( _queues_lock );
+    _frame_queue.clear();
+    _metadata_queue.clear();
+}
+
+
 void dds_metadata_syncer::enqueue_frame( key_type id, frame_holder && frame )
 {
+    std::weak_ptr< bool > alive = _is_alive;
+    if( ! alive.lock() || ! *_is_alive ) // Check if was or being destructed by other thread
```

Comment:
Technically, you could pass this test and then, before the lock, the syncer would get destroyed before we enter the lock. Right?

Reply:
Theoretically yes, but since locking is the next line it is very unlikely. Also I am resetting `on_data_available` before clearing the map so no one should be calling `enqueue_frame/metadata`.

The problems I encountered were when we returned from a callback to an already destructed syncer.